### ISSUE TITLE
Xarray NetCDFDataSet

### ISSUE THREAD

**Issue by lucasjamar** [kedro-org/kedro#1346]
## Description
Read and write netcdf file into Xarray.
https://xarray.pydata.org/en/stable/user-guide/io.html#netcdf

## Context
Should attract the weather data science community to use kedro :)

## Possible Implementation
Should be quite similar to pandas.CSVDataSet. I'll give this implementation a shot in my free time.

Has anyone ever implemented such a custom dataset already?

**Comment by datajoely** [comment]
@lucasjamar to my knowledge this hasn't been requested before - we'd absolutely appreciate a PR :)

**Comment by AntonyMilneQB** [comment]
Always interesting to hear about communities I'm not familiar with and how they might use kedro 🙂 🎉 👍 FYI @noklam - am I right in thinking you're from geoscience originally?

**Comment by noklam** [comment]
@AntonyMilneQB Close enough! I did Earth System Science.

**Comment by noklam** [comment]
@lucasjamar Hi Lucas, just checking in to see if you need any help with this.

**Comment by lucasjamar** [comment]
Hi @noklam,

Quite busy right now sorry. Im hoping to have a look at this over easter

**Comment by noklam** [comment]
@lucasjamar  No worries! Looking forward to see your PR😀.

**Comment by noklam** [comment]
@lucasjamar Are you still interested in this?

**Comment by lucasjamar** [comment]
@noklam I started during the holiday break but didn't get very far. Im afraid i wont find the time to get any further with this... Terribly sorry

**Comment by lucasjamar** [comment]
Sorry this is as far as i could get:

```
# Copyright 2021 QuantumBlack Visual Analytics Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND
# NONINFRINGEMENT. IN NO EVENT WILL THE LICENSOR OR OTHER CONTRIBUTORS
# BE LIABLE FOR ANY CLAIM, DAMAGES, OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF, OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# The QuantumBlack Visual Analytics Limited ("QuantumBlack") name and logo
# (either separately or in combination, "QuantumBlack Trademarks") are
# trademarks of QuantumBlack. The License does not grant you any right or
# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
# Trademarks or any confusingly similar mark as a trademark for your product,
# or use the QuantumBlack Trademarks in any other manner that might cause
# confusion in the marketplace, including but not limited to in advertising,
# on websites, or on software.
#
# See the License for the specific language governing permissions and
# limitations under the License.


"""NetworkX ``GMLDataSet`` loads and saves graphs to a graph modelling language (GML)
file using an underlying filesystem (e.g.: local, S3, GCS). ``NetworkX`` is used to
create GML data.
"""

from copy import deepcopy
from pathlib import PurePosixPath
from typing import Any, Dict

import fsspec
import networkx

from kedro.io.core import (
    AbstractVersionedDataSet,
    Version,
    get_filepath_str,
    get_protocol_and_path,
)


class GMLDataSet(AbstractVersionedDataSet):
    """``GMLDataSet`` loads and saves graphs to a GML file using an
    underlying filesystem (e.g.: local, S3, GCS). ``NetworkX`` is used to
    create GML data.
    See https://networkx.org/documentation/stable/tutorial.html for details.

    Example:
    ::

        >>> from kedro.extras.datasets.networkx import GMLDataSet
        >>> import networkx as nx
        >>> graph = nx.complete_graph(100)
        >>> graph_dataset = GMLDataSet(filepath="test.gml")
        >>> graph_dataset.save(graph)
        >>> reloaded = graph_dataset.load()
        >>> assert nx.is_isomorphic(graph, reloaded)

    """

    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
    DEFAULT_SAVE_ARGS = {}  # type: Dict[str, Any]

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        filepath: str,
        load_args: Dict[str, Any] = None,
        save_args: Dict[str, Any] = None,
        version: Version = None,
        credentials: Dict[str, Any] = None,
        fs_args: Dict[str, Any] = None,
    ) -> None:
        """Creates a new instance of ``GMLDataSet``.

        Args:
            filepath: Filepath in POSIX format to the NetworkX GML file.
            load_args: Arguments passed on to ``networkx.read_gml``.
                See the details in
                https://networkx.org/documentation/stable/reference/readwrite/generated/networkx.readwrite.gml.read_gml.html
            save_args: Arguments passed on to ``networkx.write_gml``.
                See the details in
                https://networkx.org/documentation/stable/reference/readwrite/generated/networkx.readwrite.gml.write_gml.html
            version: If specified, should be an instance of
                ``kedro.io.core.Version``. If its ``load`` attribute is
                None, the latest version will be loaded. If its ``save``
                attribute is None, save version will be autogenerated.
            credentials: Credentials required to get access to the underlying filesystem.
                E.g. for ``GCSFileSystem`` it should look like `{"token": None}`.
            fs_args: Extra arguments to pass into underlying filesystem class constructor
                (e.g. `{"project": "my-project"}` for ``GCSFileSystem``), as well as
                to pass to the filesystem's `open` method through nested keys
                `open_args_load` and `open_args_save`.
                Here you can find all available arguments for `open`:
                https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.open
                All defaults are preserved, except `mode`, which is set to `r` when loading
                and to `w` when saving.
        """
        _fs_args = deepcopy(fs_args) or {}
        _fs_open_args_load = _fs_args.pop("open_args_load", {})
        _fs_open_args_save = _fs_args.pop("open_args_save", {})
        _credentials = deepcopy(credentials) or {}

        protocol, path = get_protocol_and_path(filepath, version)
        if protocol == "file":
            _fs_args.setdefault("auto_mkdir", True)

        self._protocol = protocol
        self._fs = fsspec.filesystem(self._protocol, **_credentials, **_fs_args)

        super().__init__(
            filepath=PurePosixPath(path),
            version=version,
            exists_function=self._fs.exists,
            glob_function=self._fs.glob,
        )

        # Handle default load and save arguments
        self._load_args = deepcopy(self.DEFAULT_LOAD_ARGS)
        if load_args is not None:
            self._load_args.update(load_args)
        self._save_args = deepcopy(self.DEFAULT_SAVE_ARGS)
        if save_args is not None:
            self._save_args.update(save_args)
        _fs_open_args_load.setdefault("mode", "rb")
        _fs_open_args_save.setdefault("mode", "wb")
        self._fs_open_args_load = _fs_open_args_load
        self._fs_open_args_save = _fs_open_args_save

    def _load(self) -> networkx.Graph:
        load_path = get_filepath_str(self._get_load_path(), self._protocol)
        with self._fs.open(load_path, **self._fs_open_args_load) as fs_file:
            data = networkx.read_gml(fs_file, **self._load_args)
        return data

    def _save(self, data: networkx.Graph) -> None:
        save_path = get_filepath_str(self._get_save_path(), self._protocol)
        with self._fs.open(save_path, **self._fs_open_args_save) as fs_file:
            networkx.write_gml(data, fs_file, **self._save_args)
        self._invalidate_cache()

    def _exists(self) -> bool:
        load_path = get_filepath_str(self._get_load_path(), self._protocol)
        return self._fs.exists(load_path)

    def _describe(self) -> Dict[str, Any]:
        return dict(
            filepath=self._filepath,
            protocol=self._protocol,
            load_args=self._load_args,
            save_args=self._save_args,
            version=self._version,
        )

    def _release(self) -> None:
        super()._release()
        self._invalidate_cache()

    def _invalidate_cache(self) -> None:
        """Invalidate underlying filesystem caches."""
        filepath = get_filepath_str(self._filepath, self._protocol)
        self._fs.invalidate_cache(filepath)
```

**Comment by noklam** [comment]
@lucasjamar Hey thanks for making time for this! Do you want to make this a PR? It would be easier for me to drop comments on it. :)

**Comment by jamespolly** [comment]
Xarray functionality in the Data Catalog would be a big deal for me. Is there an update on this?

**Comment by noklam** [comment]
At a glance making just NetCDF works is straight forward, but seems @lucasjamar is going with a GenericXarrayDataSet. I wonder if it makes sense to take the `pandas` approach with `xarray.NetCDFDataSet`, `xarray.PickleDataSet` etc.

Here is a quick implementation, mostly just copy from above

```python

class NetCDFDataSet(AbstractVersionedDataSet):
    """``GenericDataSet`` loads/saves data from/to a file using an underlying
    filesystem (e.g.: local, S3, GCS). It uses xarray to handle the file.

    Example adding a catalog entry with
    `YAML API
    <https://kedro.readthedocs.io/en/stable/data/\
        data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:

    .. code-block:: yaml

        >>> january:
        >>>   type: xarray.GenericDataSet
        >>>   filepath: data/01_raw/weather/january.nc
        >>>   load_args:
        >>>     engine: netcdf4
        >>>     decode_times: True
        >>>   save_args:
        >>>     index: False
        >>>     date_format: "%Y-%m-%d %H:%M"
        >>>
        >>> motorbikes:
        >>>   type: xarray.GenericDataSet
        >>>   filepath: gcs://your_bucket/weater.ya
        >>>   credentials: dev_s3
        >>>


    Example using Python API:
    ::

        >>> from kedro.extras.datasets.xarray import GenericDataSet
        >>> import numpy as np
        >>> import pandas as pd
        >>> import xarray as xr
        >>>
        >>> data = xr.Dataset(
        >>>     {"foo": (("x", "y"), np.random.rand(4, 5))},
        >>>     coords={
        >>>         "x": [10, 20, 30, 40],
        >>>         "y": pd.date_range("2000-01-01", periods=5),
        >>>         "z": ("x", list("abcd")),
        >>>     },
        >>> )
        >>>
        >>> # data_set = GenericDataSet(filepath="gcs://bucket/test.nc")
        >>> data_set = GenericDataSet(filepath="test.nc")
        >>> data_set.save(data)
        >>> reloaded = data_set.load()
        >>> assert data.equals(reloaded)

    """

    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
    DEFAULT_SAVE_ARGS = {}  # type: Dict[str, Any]

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        filepath: str,
        load_args: Dict[str, Any] = None,
        save_args: Dict[str, Any] = None,
        version: Version = None,
        credentials: Dict[str, Any] = None,
        fs_args: Dict[str, Any] = None,
    ) -> None:
        """Creates a new instance of ``GenericDataSet`` pointing to a concrete NetCDF file
        on a specific filesystem.

        Args:
            filepath: Filepath in POSIX format to a NetCDF file prefixed with a protocol like `s3://`.
                If prefix is not provided, `file` protocol (local filesystem) will be used.
                The prefix should be any protocol supported by ``fsspec``.
                Note: `http(s)` doesn't support versioning.
            load_args: Pandas options for loading NetCDF files.
                Here you can find all available arguments:
                https://xarray.pydata.org/xarray-docs/stable/generated/xarray.read_nc.html
                All defaults are preserved.
            save_args: Pandas options for saving NetCDF files.
                Here you can find all available arguments:
                https://xarray.pydata.org/xarray-docs/stable/generated/xarray.DataFrame.to_nc.html
                All defaults are preserved, but "index", which is set to False.
            version: If specified, should be an instance of
                ``kedro.io.core.Version``. If its ``load`` attribute is
                None, the latest version will be loaded. If its ``save``
                attribute is None, save version will be autogenerated.
            credentials: Credentials required to get access to the underlying filesystem.
                E.g. for ``GCSFileSystem`` it should look like `{"token": None}`.
            fs_args: Extra arguments to pass into underlying filesystem class constructor
                (e.g. `{"project": "my-project"}` for ``GCSFileSystem``).
        """
        _fs_args = deepcopy(fs_args) or {}
        _fs_open_args_load = _fs_args.pop("open_args_load", {})
        _fs_open_args_save = _fs_args.pop("open_args_save", {})
        _credentials = deepcopy(credentials) or {}

        protocol, path = get_protocol_and_path(filepath, version)
        if protocol == "file":
            _fs_args.setdefault("auto_mkdir", True)

        self._protocol = protocol
        self._storage_options = {**_credentials, **_fs_args}
        self._fs = fsspec.filesystem(self._protocol, **self._storage_options)

        super().__init__(
            filepath=PurePosixPath(path),
            version=version,
            exists_function=self._fs.exists,
            glob_function=self._fs.glob,
        )

        # Handle default load and save arguments
        self._load_args = deepcopy(self.DEFAULT_LOAD_ARGS)
        if load_args is not None:
            self._load_args.update(load_args)
        self._save_args = deepcopy(self.DEFAULT_SAVE_ARGS)
        if save_args is not None:
            self._save_args.update(save_args)

        if "storage_options" in self._save_args or "storage_options" in self._load_args:
            logger.warning(
                "Dropping `storage_options` for %s, "
                "please specify them under `fs_args` or `credentials`.",
                self._filepath,
            )
            self._save_args.pop("storage_options", None)
            self._load_args.pop("storage_options", None)

        _fs_open_args_save.setdefault("mode", "wb")
        self._fs_open_args_load = _fs_open_args_load
        self._fs_open_args_save = _fs_open_args_save

    def _describe(self) -> Dict[str, Any]:
        return dict(
            filepath=self._filepath,
            protocol=self._protocol,
            load_args=self._load_args,
            save_args=self._save_args,
            version=self._version,
        )

    def _load(self) -> xr.Dataset:
        load_path = str(self._get_load_path())
        if self._protocol == "file":
            # file:// protocol seems to misbehave on Windows
            # (<urlopen error file not on local host>),
            # so we don't join that back to the filepath;
            # storage_options also don't work with local paths
            return xr.open_dataset(load_path, **self._load_args)

        load_path = f"{self._protocol}{PROTOCOL_DELIMITER}{load_path}"
        return xr.open_dataset(
            load_path, storage_options=self._storage_options, **self._load_args
        )

    def _save(self, data: xr.DataArray) -> None:
        save_path = get_filepath_str(self._get_save_path(), self._protocol)
        with self._fs.open(save_path, **self._fs_open_args_save) as fs_file:
            data.to_netcdf(fs_file, **self._save_args)

        self._invalidate_cache()

    def _exists(self) -> bool:
        try:
            load_path = get_filepath_str(self._get_load_path(), self._protocol)
        except DataSetError:
            return False

        return self._fs.exists(load_path)

    def _release(self) -> None:
        super()._release()
        self._invalidate_cache()

    def _invalidate_cache(self) -> None:
        """Invalidate underlying filesystem caches."""
        filepath = get_filepath_str(self._filepath, self._protocol)
        self._fs.invalidate_cache(filepath)
```

**Comment by AntonyMilneQB** [comment]
Would it make sense to have a `file_format` argument and then get the right `xr.to/from_***` function based on that?I'm not sure how standardised the xarray load/save API is. This is what we do in `pandas.GenericDataSet`:
https://github.com/kedro-org/kedro/blob/fdc6be73408852b09fcc16370706257b918994d5/kedro/extras/datasets/pandas/generic_dataset.py#L187-L188

**Comment by noklam** [comment]
I am actually not aware of the pandas.GenericDataSet😅 do people actually use that and do we keep both for backward compatiblitity reason? I think it is possible since the API is quite similar.

Btw I saw the invalidate_cache method for some datasets are public but the majority are private, does that make sense if we just remove that and put it in the abstractClass? We have repeated that 71 times😅

**Comment by AntonyMilneQB** [comment]
Yeah, `pandas.GenericDataSet` isn't well known. It was added relatively recently so that we don't have to constantly be catching up with pandas when they add a new format: https://github.com/kedro-org/kedro/pull/987. It's not really designed to replace the `pandas.CSVDataSet` etc. - just to cater for any new formats where we don't have a dedicated dataset. But at some point it might make sense to deprecate the dedicated ones in favour of the generic.

> Btw I saw the invalidate_cache method for some datasets are public but the majority are private, does that make sense if we just remove that and put it in the abstractClass? We have repeated that 71 times😅

Probably yes. The same could probably be said for `_exists`, `_release`, the repetition of `save_path = get_filepath_str(self._filepath, self._protocol)`, etc. etc. It would be nice if we could reduce the repeated boiler plate that's in every dataset definition, and I don't think it would be too hard to do. Possibly there's some reason for keeping it, so let's not remove it straight away without getting some more input. I would consider it part of my "fundamental issue" in https://github.com/kedro-org/kedro/issues/1691 - we should consider generally what exactly belongs in the abstract classes vs. the implementations.

**Comment by noklam** [comment]
Not sure if I like the idea of deprecating the dedicated one, but this is not urgent so I will think more about it. I added the `dataset` related refactoring piece into #1691 in your comments, it's a bit easier to keep track of when we have a dedicated issue for that.

**Comment by datajoely** [comment]
Yeah the main reason why I pushed for `pandas.GenericDataSet` is because they introduced `pd.read_xml` and it felt like we were just always in reactive mode and could at least provide a generic way of users leveraging pandas features without us needing to push a release.

The pandas API is also really in consistently, for file paths I had to rely on the assumption that the first positional argument is going to be the path argument since the library uses a different `kwarg` because each reader/writer was contributed independently. 

In the Spark world we're able to only provide a generic version of the dataset since it's much better designed, I think in situations like `xarray` and `dask` we can probably do the same.

**Comment by jamespolly** [comment]
I really appreciate the responses. :pray: 

Re: the "pandas approach" mentioned by @noklam, I think there are [certainly a few xarray-supported file types](https://docs.xarray.dev/en/stable/user-guide/io.html) that could benefit from built-in handling in kedro. NetCDF, Zarr, and GeoTIFF (via [rioxarray](https://corteva.github.io/rioxarray/stable/getting_started/getting_started.html#rioxarray)) specifically. Maybe that supports moving away from a generic dataset, but I also appreciate the difficulty in supporting everything.

In the selfish short term it looks like I can use the implementation provided by @noklam above in combination with [these instructions](https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html) to get done what I need to get done. I can test out these implementations and report back changes I see as necessary.

**Comment by AntonyMilneQB** [comment]
Definitely that sounds like a good plan in the short term - you can easily use the custom dataset implementation without it being added to core kedro. Just copy and paste the above into a file and modify the dataset `type` to point to that class. This would be a great way to test it out and see how well it all works.

**Comment by jamespolly** [comment]
An update and a couple questions about partitioned datasets and lazy evaluation. The implementation discussed above works well. I've only tested it on local file systems.

TL;DR: To not lose advantages of working with xarray we need to concatenate many .nc files into a big dataset and then operate on it as lazily as possible. I want to keep benefits of both xarray and Kedro.
1. When saving data to the Data Catalog, is the `create_partitions` approach I'm employing here best/as-intended by Kedro? I'm using a dictionary of `{ids: xarray.DataSets}`.
2. When not saving to the Data Catalog and passing node outputs ("in memory") to the inputs of the next node, is this being done lazily or are things being loaded? Dask is the backend for xarray if this matters, and what Kedro means by "in memory" starts to get fuzzy for me with data types supporting lazy eval.

More on this:
When obtaining NetCDF files from a remote source, which are being saved for the first time in a pipeline (e.g., saving to `/data/01_raw`), it is common (outside of Kedro) to do this with a for loop, saving every iteration. With Kedro handling file I/O in a partitioned dataset, is it best to make a function that builds a dictionary of unique partition ID keys with corresponding xarray.DataSet values? 

This approach (example shown below) works, but I'm not sure if it is loading all this data into memory and ruining the lazy evaluation aspect. The only way I know to test this is to `htop` while increasing the number of datasets I'm downloading to exceed the memory of my machine. Hoping to get some feedback and learn before I do that.

In this pipeline I would have a node defined in `pipeline.py` as:
```
node(
    func=create_data_partitions,
    inputs=["params:starttime", "params:endtime"],
    outputs="downloaded_data", # defined in catalog.yml
    name="download_forecast_node",
    )
```
With the corresponding function definition (which as shown here is a very simplified/broken example to highlight dictionary creation aspect) in `nodes.py` as:

```
def create_data_partitions(starttime: Dict, endtime: Dict) -> Dict[str, Any]:
    """Create new partitions and save using PartitionedDataSet.

    returns: Dictionary of the partitions to create.
    """
    duration = endtime - starttime
    part_lkup = {}
    for i in range(duration):
        tmptime = starttime + timedelta(hours=i)
        ds = get_data(tmptime) # downloads a single xarray dataset
        part_id = str(tmptime)+'_downloaded'
        part_lkup[part_id] = ds

    return part_lkup
```

**I've used this pattern in multiple places, and it works, but again it's not clear to me that it is the best way to go.**

Another example, when I start a processing pipeline on this downloaded data, I'll open the downloaded data files but concatenate them into a single, large (but not loaded into memory) xarray.DataSet. In doing so (as shown below) I am attempting to do what [xr.open_mfdataset](https://docs.xarray.dev/en/stable/generated/xarray.open_mfdataset.html) will do when given a directory full of `.nc` files.

```
def create_pipeline(**kwargs) -> Pipeline:
    return pipeline(
        [
            node(
                func=create_data_partitions,
                inputs=["params:starttime", "params:endtime"],
                outputs="downloaded_data", # defined in catalog.yml
                name="download_forecast_node",
            ),
            node(
                func=concat_partitions,
                inputs="downloaded_data",
                outputs="concat_ds", # not in catalog, in memory, hopefully still lazy and not actually loaded
                name="concat_download_ds_node",
            ),
            node(
                func=cleanup_data,
                inputs="concat_ds",
                outputs="clean_ds", # not in catalog, in memory, hopefully still lazy and not actually loaded
                name="clean_ds_node",
            ),
            node(
                func=create_partitions,
                inputs="clean_ds",
                outputs="forecast_clean",
                name="forecast_clean_node", # defined in catalog.yml
            ),
        ]
    )
```
Where the `concat_partitions` node uses a [pattern similar to those in the Kedro docs](https://kedro.readthedocs.io/en/stable/data/kedro_io.html#partitioned-dataset-load):

```
def concat_partitions(partitioned_input: Dict[str, Callable[[], Any]], concat_dim: str = 'time') -> xr.Dataset:
    """Concatenate input partitions into one xarray Dataset.

    param: partitioned_input: A dictionary with partition ids as keys and load functions as values.
    param: concat_dim: existing dimension in each netcdf file along which to concatenate all datasets.

    return: xarray Dataset containing all loaded partitions.
    """

    return xr.concat([partitioned_input[i]() for i in sorted(partitioned_input)], dim=concat_dim)
```

and where the `create_partitions` node uses my same uncertain/questionable pattern at the heart of this question:
```
def create_partitions(ds:xr.Dataset) -> Dict[str, Any]:
    """Create new partitions and save using PartitionedDataSet.

    returns: Dictionary of the partitions to create.
    """
    times = ds.time.values
    part_lkup = {}
    for tmptime in times:
        tmp_ds = ds.sel(time=slice(tmptime, tmptime))
        part_id = str(tmptime)[:13]+'_cleaned_data'
        part_lkup[part_id] = tmp_ds

    return part_lkup
```

**Comment by datajoely** [comment]
@jamespolly thank you for your very detailed write up - @deepyaman do you want to chime in with your thoughts?

**Comment by NowanIlfideme** [comment]
One issue I've run into with `xarray.load_dataset()` and `ds.to_netcdf()` functions is that they still don't allow reading/writing from bytes, only from local files. Would be nice if I could just use Kedro to abstract this away. 😄
